File: src/test/java/cube8540/oauth/authentication/credentials/oauth/client/application/DefaultOAuth2ClientManagementServiceTest.java
package cube8540.oauth.authentication.credentials.oauth.client.application;

import cube8540.oauth.authentication.credentials.oauth.client.domain.OAuth2Client;
import cube8540.oauth.authentication.credentials.oauth.client.domain.OAuth2ClientRepository;
import cube8540.oauth.authentication.credentials.oauth.client.domain.OAuth2ClientValidatorFactory;
import cube8540.oauth.authentication.credentials.oauth.client.domain.exception.ClientErrorCodes;
import cube8540.oauth.authentication.credentials.oauth.client.domain.exception.ClientNotFoundException;
import cube8540.oauth.authentication.credentials.oauth.client.domain.exception.ClientRegisterException;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.InOrder;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;

import java.time.Duration;

import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.ACCESS_TOKEN_VALIDITY_SECONDS;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.CLIENT_ID;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.CLIENT_NAME;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.ENCODING_SECRET;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.GRANT_TYPES;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.MODIFY_ACCESS_TOKEN_VALIDITY_SECONDS;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.MODIFY_CLIENT_NAME;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.MODIFY_REFRESH_TOKEN_VALIDITY_SECONDS;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.MODIFY_SECRET;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.NEW_GRANT_TYPES;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.NEW_REDIRECT_URIS;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.NEW_SCOPES;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.OWNER;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.RAW_CLIENT_ID;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.RAW_GRANT_TYPES;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.RAW_NEW_GRANT_TYPES;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.RAW_NEW_REDIRECT_URIS;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.RAW_NEW_SCOPES;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.RAW_OWNER;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.RAW_REDIRECT_URIS;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.RAW_REMOVE_GRANT_TYPES;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.RAW_REMOVE_REDIRECT_URIS;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.RAW_REMOVE_SCOPES;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.RAW_SCOPES;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.REDIRECT_URIS;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.REFRESH_TOKEN_VALIDITY_SECONDS;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.REMOVE_GRANT_TYPES;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.REMOVE_REDIRECT_URIS;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.REMOVE_SCOPES;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.SCOPES;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.SECRET;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.makeClientRepository;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.makeDefaultClient;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.makeEmptyClientRepository;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.makeEncoder;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.makeErrorValidatorFactory;
import static cube8540.oauth.authentication.credentials.oauth.client.application.OAuth2ClientApplicationTestHelper.makeValidatorFactory;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.inOrder;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

@DisplayName("기본 클라이언트 관리 서비스 테스트")
class DefaultOAuth2ClientManagementServiceTest {

    @Test
    @DisplayName("이미 등록 되어 있는 클라이언트 등록")
    void registerAlreadyRegisteredClientInRepository() {
        OAuth2Client client = makeDefaultClient();
        OAuth2ClientRepository repository = makeClientRepository(CLIENT_ID, client);
        OAuth2ClientRegisterRequest request = new OAuth2ClientRegisterRequest(RAW_CLIENT_ID, SECRET, CLIENT_NAME,
                RAW_REMOVE_REDIRECT_URIS, RAW_SCOPES, RAW_GRANT_TYPES, ACCESS_TOKEN_VALIDITY_SECONDS, REFRESH_TOKEN_VALIDITY_SECONDS, RAW_OWNER);
        DefaultOAuth2ClientManagementService service = new DefaultOAuth2ClientManagementService(repository);

        ClientRegisterException e = assertThrows(ClientRegisterException.class, () -> service.registerNewClient(request));
        Assertions.assertEquals(ClientErrorCodes.EXISTS_IDENTIFIER, e.getCode());
    }

    @Test
    @DisplayName("허용 되지 않는 클라이언트 등록")
    void registerNewClientWithInvalidData() {
        OAuth2ClientRegisterRequest request = new OAuth2ClientRegisterRequest(RAW_CLIENT_ID, SECRET, CLIENT_NAME,
                RAW_REDIRECT_URIS, RAW_SCOPES, RAW_GRANT_TYPES, ACCESS_TOKEN_VALIDITY_SECONDS, REFRESH_TOKEN_VALIDITY_SECONDS, RAW_OWNER);
        PasswordEncoder encoder = makeEncoder(SECRET, ENCODING_SECRET);
        OAuth2ClientRepository repository = makeEmptyClientRepository();
        OAuth2ClientValidatorFactory factory = makeErrorValidatorFactory(new TestOauth2ClientException());
        DefaultOAuth2ClientManagementService service = new DefaultOAuth2ClientManagementService(repository);

        service.setValidateFactory(factory);
        service.setPasswordEncoder(encoder);

        assertThrows(TestOauth2ClientException.class, () -> service.registerNewClient(request));
    }

    @Test
    @DisplayName("새 클라이언트 등록")
    void registerNewClient() {
        ArgumentCaptor<OAuth2Client> clientCaptor = ArgumentCaptor.forClass(OAuth2Client.class);
        OAuth2ClientRegisterRequest request = new OAuth2ClientRegisterRequest(RAW_CLIENT_ID, SECRET, CLIENT_NAME,
                RAW_REDIRECT_URIS, RAW_SCOPES, RAW_GRANT_TYPES, ACCESS_TOKEN_VALIDITY_SECONDS, REFRESH_TOKEN_VALIDITY_SECONDS, RAW_OWNER);
        PasswordEncoder encoder = makeEncoder(SECRET, ENCODING_SECRET);
        OAuth2ClientRepository repository = makeEmptyClientRepository();
        OAuth2ClientValidatorFactory factory = makeValidatorFactory();
        DefaultOAuth2ClientManagementService service = new DefaultOAuth2ClientManagementService(repository);

        service.setValidateFactory(factory);
        service.setPasswordEncoder(encoder);

        service.registerNewClient(request);
        verify(repository, times(1)).save(clientCaptor.capture());
        assertEquals(CLIENT_ID, clientCaptor.getValue().getClientId());
        assertEquals(ENCODING_SECRET, clientCaptor.getValue().getSecret());
        assertEquals(OWNER, clientCaptor.getValue().getOwner());
        assertEquals(REDIRECT_URIS, clientCaptor.getValue().getRedirectUris());
        assertEquals(SCOPES, clientCaptor.getValue().getScopes());
        assertEquals(GRANT_TYPES, clientCaptor.getValue().getGrantTypes());
        assertEquals(Duration.ofSeconds(ACCESS_TOKEN_VALIDITY_SECONDS), clientCaptor.getValue().getAccessTokenValidity());
        assertEquals(Duration.ofSeconds(REFRESH_TOKEN_VALIDITY_SECONDS), clientCaptor.getValue().getRefreshTokenValidity());
    }

    @Test
    @DisplayName("저장소에 등록 되어 있지 않은 클라이언트 수정")
    void modifyNotRegisteredClientInRepository() {
        OAuth2ClientModifyRequest request = new OAuth2ClientModifyRequest(MODIFY_CLIENT_NAME, RAW_NEW_REDIRECT_URIS,
                RAW_REMOVE_REDIRECT_URIS, RAW_NEW_GRANT_TYPES, RAW_REMOVE_GRANT_TYPES, RAW_NEW_SCOPES, RAW_REMOVE_SCOPES,
                MODIFY_ACCESS_TOKEN_VALIDITY_SECONDS, MODIFY_REFRESH_TOKEN_VALIDITY_SECONDS);
        OAuth2ClientRepository repository = makeEmptyClientRepository();
        DefaultOAuth2ClientManagementService service = new DefaultOAuth2ClientManagementService(repository);

        assertThrows(ClientNotFoundException.class, () -> service.modifyClient(OAuth2ClientApplicationTestHelper.RAW_CLIENT_ID, request));
    }

    @Test
    @DisplayName("클라이언트 수정")
    void modifyClient() {
        OAuth2Client client = makeDefaultClient();
        OAuth2ClientModifyRequest request = new OAuth2ClientModifyRequest(MODIFY_CLIENT_NAME, RAW_NEW_REDIRECT_URIS,
                RAW_REMOVE_REDIRECT_URIS, RAW_NEW_GRANT_TYPES, RAW_REMOVE_GRANT_TYPES, RAW_NEW_SCOPES, RAW_REMOVE_SCOPES,
                MODIFY_ACCESS_TOKEN_VALIDITY_SECONDS, MODIFY_REFRESH_TOKEN_VALIDITY_SECONDS);
        OAuth2ClientRepository repository = makeClientRepository(CLIENT_ID, client);
        OAuth2ClientValidatorFactory factory = makeValidatorFactory();
        DefaultOAuth2ClientManagementService service = new DefaultOAuth2ClientManagementService(repository);

        service.setValidateFactory(factory);

        InOrder inOrder = inOrder(client, repository);
        service.modifyClient(OAuth2ClientApplicationTestHelper.RAW_CLIENT_ID, request);
        inOrder.verify(client, times(1)).setClientName(MODIFY_CLIENT_NAME);
        REMOVE_REDIRECT_URIS.forEach(uri -> inOrder.verify(client, times(1)).removeRedirectUri(uri));
        NEW_REDIRECT_URIS.forEach(uri -> inOrder.verify(client, times(1)).addRedirectUri(uri));
        REMOVE_GRANT_TYPES.forEach(grant -> inOrder.verify(client, times(1)).removeGrantType(grant));
        NEW_GRANT_TYPES.forEach(grant -> inOrder.verify(client, times(1)).addGrantType(grant));
        REMOVE_SCOPES.forEach(scope -> inOrder.verify(client, times(1)).removeScope(scope));
        NEW_SCOPES.forEach(scope -> inOrder.verify(client, times(1)).addScope(scope));
        inOrder.verify(client, times(1)).setAccessTokenValidity(MODIFY_ACCESS_TOKEN_VALIDITY_SECONDS);
        inOrder.verify(client, times(1)).setRefreshTokenValidity(MODIFY_REFRESH_TOKEN_VALIDITY_SECONDS);
        inOrder.verify(client, times(1)).validate(factory);
        inOrder.verify(repository, times(1)).save(client);
    }

    @Test
    @DisplayName("저장소에 등록 되어 있지 않은 클라이언트 패스워드 변경")
    void changePasswordNotRegisteredClientInRepository() {
        OAuth2ChangeSecretRequest request = new OAuth2ChangeSecretRequest(SECRET, MODIFY_SECRET);
        OAuth2ClientRepository repository = makeEmptyClientRepository();
        DefaultOAuth2ClientManagementService service = new DefaultOAuth2ClientManagementService(repository);

        assertThrows(ClientNotFoundException.class, () -> service.changeSecret(RAW_CLIENT_ID, request));
    }

    @Test
    @DisplayName("클라이언트 패스워드 변경")
    void changeClientSecret() {
        OAuth2ChangeSecretRequest request = new OAuth2ChangeSecretRequest(SECRET, MODIFY_SECRET);
        OAuth2Client client = makeDefaultClient();
        OAuth2ClientRepository repository = makeClientRepository(CLIENT_ID, client);
        PasswordEncoder encoder = makeEncoder(MODIFY_SECRET, ENCODING_SECRET);
        OAuth2ClientValidatorFactory factory = makeValidatorFactory();
        DefaultOAuth2ClientManagementService service = new DefaultOAuth2ClientManagementService(repository);

        service.setValidateFactory(factory);
        service.setPasswordEncoder(encoder);

        InOrder inOrder = inOrder(client, repository);
        service.changeSecret(RAW_CLIENT_ID, request);
        inOrder.verify(client, times(1)).changeSecret(SECRET, MODIFY_SECRET, encoder);
        inOrder.verify(client, times(1)).validate(factory);
        inOrder.verify(client, times(1)).encrypted(encoder);
        inOrder.verify(repository, times(1)).save(client);
    }

    @Test
    @DisplayName("저장소에 등록 되어 있지 않은 클라이언트 삭제")
    void removeNotRegisteredClientInRepository() {
        OAuth2ClientRepository repository = makeEmptyClientRepository();
        DefaultOAuth2ClientManagementService service = new DefaultOAuth2ClientManagementService(repository);

        assertThrows(ClientNotFoundException.class, () -> service.removeClient(RAW_CLIENT_ID));
    }

    @Test
    @DisplayName("클라이언트 삭제")
    void removeClient() {
        OAuth2Client client = makeDefaultClient();
        OAuth2ClientRepository repository = makeClientRepository(CLIENT_ID, client);
        DefaultOAuth2ClientManagementService service = new DefaultOAuth2ClientManagementService(repository);

        service.removeClient(RAW_CLIENT_ID);
        verify(repository, times(1)).delete(client);
    }

    @AfterEach
    void after() {
        SecurityContextHolder.clearContext();
    }

    private static class TestOauth2ClientException extends RuntimeException {}
}

File: src/test/kotlin/cube8540/oauth/authentication/oauth/client/application/DefaultOAuth2ClientManagementServiceTest.kt
package cube8540.oauth.authentication.oauth.client.application

import cube8540.oauth.authentication.UnitTestValidationException
import cube8540.oauth.authentication.error.message.ErrorCodes
import cube8540.oauth.authentication.oauth.client.domain.*
import cube8540.oauth.authentication.oauth.extractGrantType
import cube8540.oauth.authentication.security.AuthorityCode
import io.mockk.*
import java.net.URI
import java.time.Duration
import java.util.Optional
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.catchThrowable
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.oauth2.core.AuthorizationGrantType

class DefaultOAuth2ClientManagementServiceTest {

    private val repository: OAuth2ClientRepository = mockk()
    private val passwordEncoder: PasswordEncoder = mockk()
    private val validatorFactory: OAuth2ClientValidatorFactory = mockk()

    private val service = DefaultOAuth2ClientManagementService(repository)

    init {
        service.validateFactory = validatorFactory
        service.passwordEncoder = passwordEncoder
    }

    @Nested
    inner class ClientRegisterTest {

        @Test
        fun `client already registered`() {
            val registerRequest = OAuth2ClientRegisterRequest(clientId = "clientId", secret = "clientSecret", clientName = "clientName",
                redirectUris = emptyList(), scopes = emptyList(), grantTypes = emptyList(),
                accessTokenValiditySeconds = 0, refreshTokenValiditySeconds = 0, clientOwner = "owner")

            every { repository.countByClientId(OAuth2ClientId("clientId")) } returns 1

            val thrown = catchThrowable { service.registerNewClient(registerRequest) }
            assertThat(thrown).isInstanceOf(ClientRegisterException::class.java)
            assertThat((thrown as ClientRegisterException).code).isEqualTo(ErrorCodes.EXISTS_IDENTIFIER)
        }

        @Test
        fun `register client data invalid`() {
            val registerRequest = OAuth2ClientRegisterRequest(clientId = "clientId", secret = "clientSecret", clientName = "clientName",
                redirectUris = emptyList(), scopes = emptyList(), grantTypes = emptyList(),
                accessTokenValiditySeconds = 0, refreshTokenValiditySeconds = 0, clientOwner = "owner")

            every { repository.countByClientId(OAuth2ClientId("clientId")) } returns 0
            every { validatorFactory.createValidator(any()) } returns mockk {
                every { result } returns mockk {
                    every { hasErrorThrows(any()) } throws UnitTestValidationException()
                }
            }

            val thrown = catchThrowable { service.registerNewClient(registerRequest) }
            assertThat(thrown).isInstanceOf(UnitTestValidationException::class.java)
        }

        @Test
        fun `register new client`() {
            val requestRedirectUri = listOf("http://locahost/1", "http://locahost/2", "http://locahost/2")
            val requestScopes = listOf("scope-1", "scope-2", "scope-3")
            val requestGrant = listOf("authorization_code", "password", "client_credentials")
            val registerRequest = OAuth2ClientRegisterRequest(clientId = "clientId", secret = "clientSecret", clientName = "clientName",
                redirectUris = requestRedirectUri, scopes = requestScopes, grantTypes = requestGrant,
                accessTokenValiditySeconds = 10, refreshTokenValiditySeconds = 20, clientOwner = "owner")

            val clientCaptor = slot<OAuth2Client>()

            every { repository.countByClientId(OAuth2ClientId("clientId")) } returns 0
            every { repository.save(capture(clientCaptor)) } returnsArgument 0
            every { passwordEncoder.encode("clientSecret") } returns "encodedClientSecret"
            every { validatorFactory.createValidator(any()) } returns mockk {
                every { result } returns mockk {
                    every { hasErrorThrows(any()) } just Runs
                }
            }

            val expectedRedirectUrl = setOf(URI.create("http://locahost/1"), URI.create("http://locahost/2"), URI.create("http://locahost/2"))
            val expectedScope = setOf(AuthorityCode("scope-1"), AuthorityCode("scope-2"), AuthorityCode("scope-3"))
            val expectedGrantType = setOf(AuthorizationGrantType.AUTHORIZATION_CODE, AuthorizationGrantType.PASSWORD, AuthorizationGrantType.CLIENT_CREDENTIALS)
            service.registerNewClient(registerRequest)
            assertThat(clientCaptor.captured.clientId).isEqualTo(OAuth2ClientId("clientId"))
            assertThat(clientCaptor.captured.secret).isEqualTo("encodedClientSecret")
            assertThat(clientCaptor.captured.clientName).isEqualTo("clientName")
            assertThat(clientCaptor.captured.redirectUris).isEqualTo(expectedRedirectUrl)
            assertThat(clientCaptor.captured.scopes).isEqualTo(expectedScope)
            assertThat(clientCaptor.captured.grantTypes).isEqualTo(expectedGrantType)
            assertThat(clientCaptor.captured.accessTokenValidity).isEqualTo(Duration.ofSeconds(10))
            assertThat(clientCaptor.captured.refreshTokenValidity).isEqualTo(Duration.ofSeconds(20))
        }
    }

    @Nested
    inner class ClientModifyTest {

        @Test
        fun `client not registered in repository`() {
            val modifyRequest = OAuth2ClientModifyRequest(clientName = "modifyName",
                newRedirectUris = emptyList(), removeRedirectUris = emptyList(),
                newGrantTypes = emptyList(), removeGrantTypes = emptyList(),
                newScopes = emptyList(), removeScopes = emptyList(),
                accessTokenValiditySeconds = 0, refreshTokenValiditySeconds = 0
            )

            every { repository.findByClientId(OAuth2ClientId("clientId")) } returns Optional.empty()

            val thrown = catchThrowable { service.modifyClient("clientId", modifyRequest) }
            assertThat(thrown).isInstanceOf(ClientNotFoundException::class.java)
        }

        @Test
        fun `modify client`() {
            val newRedirectUris = listOf("http://localhost/new1", "http://localhost/new2", "http://localhost/new3")
            val removeRedirectUris = listOf("http://localhost/remove1", "http://localhost/remove2", "http://localhost/remove3")
            val newGrantType = listOf("authorization_code", "client_credentials")
            val removeGrantType = listOf("password")
            val newScope = listOf("new-scope-1", "new-scope-2", "new-scope-3")
            val removeScope = listOf("remove-scope-1", "remove-scope-2", "remove-scope-3")
            val modifyRequest = OAuth2ClientModifyRequest(clientName = "modifyName",
                newRedirectUris = newRedirectUris, removeRedirectUris = removeRedirectUris,
                newGrantTypes = newGrantType, removeGrantTypes = removeGrantType,
                newScopes = newScope, removeScopes = removeScope,
                accessTokenValiditySeconds = 10, refreshTokenValiditySeconds = 20)
            val client: OAuth2Client = mockk(relaxed = true)

            every { repository.findByClientId(OAuth2ClientId("clientId")) } returns Optional.of(client)
            every { repository.save(client) } returnsArgument 0

            service.modifyClient("clientId", modifyRequest)
            verifyOrder {
                client.clientName = "modifyName"
                removeRedirectUris.forEach { client.removeRedirectUri(URI.create(it)) }
                newRedirectUris.forEach { client.addRedirectUri(URI.create(it)) }

                removeGrantType.forEach { client.removeGrantType(extractGrantType(it)) }
                newGrantType.forEach { client.addGrantType(extractGrantType(it)) }

                removeScope.forEach { client.removeScope(AuthorityCode(it)) }
                newScope.forEach { client.addScope(AuthorityCode(it)) }

                client.setAccessTokenValidity(10)
                client.setRefreshTokenValidity(20)

                client.validate(validatorFactory)
                repository.save(client)
            }
        }
    }

    @Nested
    inner class ModifySecretTest {

        @Test
        fun `client not registered in repository`() {
            val changeRequest = OAuth2ChangeSecretRequest("secret", "newSecret")

            every { repository.findByClientId(OAuth2ClientId("clientId")) } returns Optional.empty()

            val thrown = catchThrowable { service.changeSecret("clientId", changeRequest) }
            assertThat(thrown).isInstanceOf(ClientNotFoundException::class.java)
        }

        @Test
        fun `change client secret`() {
            val client: OAuth2Client = mockk(relaxed = true)
            val changeRequest = OAuth2ChangeSecretRequest("secret", "newSecret")

            every { repository.findByClientId(OAuth2ClientId("clientId")) } returns Optional.of(client)
            every { repository.save(client) } returnsArgument 0

            service.changeSecret("clientId", changeRequest)
            verifyOrder {
                client.changeSecret("secret", "newSecret", passwordEncoder)
                client.validate(validatorFactory)
                client.encrypted(passwordEncoder)
                repository.save(client)
            }
        }
    }

    @Nested
    inner class RemoveTest {

        @Test
        fun `client not registered in repository`() {
            every { repository.findByClientId(OAuth2ClientId("clientId")) } returns Optional.empty()

            val thrown = catchThrowable { service.removeClient("clientId") }
            assertThat(thrown).isInstanceOf(ClientNotFoundException::class.java)
        }

        @Test
        fun `remove client`() {
            val client: OAuth2Client = mockk(relaxed = true)

            every { repository.findByClientId(OAuth2ClientId("clientId")) } returns Optional.of(client)
            every { repository.delete(client) } returnsArgument 0

            service.removeClient("clientId")
            verify { repository.delete(client) }
        }
    }

    @AfterEach
    fun clear() {
        clearAllMocks()
    }
}